### PR TITLE
[JENKINS-32183] Added disabledSince property

### PR DESCRIPTION
**Pull request by v1v**
in order to show when a job was disabled and be able to track it easily. 

Jira ticket: https://issues.jenkins-ci.org/browse/JENKINS-32183

### CODE REVIEW

File: core/src/main/java/hudson/model/AbstractProject.java [hunk 1]
```diff
@@ -688,6 +694,18 @@ public boolean isDisabled() {
     }
 
     /**
+     * When the Job is disabled.
+     *
+     * @since 1.640
```
Comment (KostyaSha) [inline core/src/main/java/hudson/model/AbstractProject.java]: since TODO

File: core/src/main/java/hudson/model/AbstractProject.java [hunk 2]
```diff
@@ -210,6 +211,11 @@
     protected volatile boolean disabled;
 
     /**
+     * When the job is disabled.
+     */
+    protected long disabledSince;
```
Comment (KostyaSha) [inline core/src/main/java/hudson/model/AbstractProject.java]: imho should be `Long` for not disabled project should have null instead of some long value.

**Comment by KostyaSha** [comment]
:-1: for lack of migration tests. Please see test examples with `@LocalData`

**Comment by v1v** [comment]
@KostyaSha thanks for your feedback

**Comment by olivergondza** [comment]
I more often wonder *why* was job disabled. IMO we should replace boolean with `DisabledCause` (quite like `OfflineCause`) that would track time, identity and optional message + other fields if needed in future.

**Comment by KostyaSha** [comment]
@olivergondza And attach to actions?

**Comment by v1v** [comment]
I was also considering that approach but I was a bit sceptical whether that feature would be quite intrusive and somehow no backward compatible. Although the disable cause might be easily fixed, as long as, it sets those details in the job description.

**Comment by v1v** [comment]
What do you want me to do? encapsulate it via a some setters/getters? 
Cheers

**Comment by olivergondza** [comment]
Ye, `getDisabledSince` and `makeDisabled` are already there to modify the state, there is no need to expose the field. FTR, I still think we need to use and object to make this easily extensible, not another scalar field. Note this approach requires synchronization to make `disabled` and `disabledSince` in sync.

File: core/src/main/java/hudson/model/AbstractProject.java [hunk 3]
```diff
@@ -711,8 +730,10 @@ public void makeDisabled(boolean b) throws IOException {
         if(disabled==b)     return; // noop
         if (b && !supportsMakeDisabled()) return; // do nothing if the disabling is unsupported
         this.disabled = b;
-        if(b)
+        if (b) {
+            this.disabledSince = System.currentTimeMillis();
             Jenkins.getInstance().getQueue().cancel(this);
+        }
```
Comment (olivergondza) [inline core/src/main/java/hudson/model/AbstractProject.java]: `disabledSince` is never reset.

**Comment by daniel-beck** [comment]
This looks a lot like the (admittedly much more stupid) recording of who created a slave: A very limited solution for a narrow requirement. More general solutions like Job Config History or something similar to Audit Trail would probably be better than this. Therefore :-1:

**Comment by KostyaSha** [comment]
@daniel-beck should be more likely a SlaveOfflineCause imho

**Comment by oleg-nenashev** [comment]
I agree with the comments above. IIRC we already had such request and decided to close it after a long discussion. BTW it would be useful to improve the widget contribution capabilities to address such "short" statuses on the main page

**Comment by v1v** [comment]
IMO, knowing when a job was disabled and who also did it in an automated manner might help to track/debug/remove those jobs easily.

I was also coding the below feature (https://github.com/v1v/jenkins/commit/1a7fe8a0c0adeb2192f77fc3a2165961c78f5052) [which it's similar to the OfflineCause feature] based on @olivergondza 's feedback, but I was struggling when showing the TextArea in the Job Dashboard.

I'll look at the Job Config History plugin and see whether that disable state it's also part of the reports. 

Cheers

**Comment by olivergondza** [comment]
@v1v, I do not think you have to struggle with the way to provide the message for now. Refactoring with the timestamp *you need* is sufficient.

**Comment by daniel-beck** [comment]
Good point with the comparison to offline causes, I'm retracting my -1. In fact, being able to enter an explanation why a job was disabled would make this significant enough to be worthwhile IMO. This could also record who did it. Knowing when/who/why seems sensible.

**Comment by oleg-nenashev** [comment]
@v1v any plans to finalize it?

**Comment by daniel-beck** [comment]
No response in a few months, so assuming this has been abandoned. Please reopen or file a new PR if you want to finish this.

**Pull request closed without merging**